relay-replays/src/recording/protocol.rs

```diff
@@ -0,0 +1,171 @@
+pub fn serialize(headers: &[u8], body: Vec<u8>) -> Result<Vec<u8>, ProtocolError> {
+    let output_bytes = compress(body)?;
+    Ok([headers.into(), vec![b'\n'], output_bytes].concat())
+}
+
+fn compress(output: Vec<u8>) -> Result<Vec<u8>, ProtocolError> {
+    let mut encoder = ZlibEncoder::new(Vec::new(), Compression::default());
+    encoder
+        .write_all(&output)
+        .map_err(ProtocolError::EncodingError)?;
+    encoder.finish().map_err(ProtocolError::EncodingError)
+}
+
+fn decompress(zipped_bytes: &[u8], limit: usize) -> Result<Vec<u8>, ProtocolError> {
+    let mut buffer = Vec::new();
+
+    let decoder = ZlibDecoder::new(zipped_bytes);
+    decoder
+        .take(limit as u64)
+        .read_to_end(&mut buffer)
+        .map_err(ProtocolError::EncodingError)?;
+
+    Ok(buffer)
+}
+
+fn read(bytes: &[u8]) -> Result<(&[u8], &[u8]), ProtocolError> {
+    match bytes.is_empty() {
+        true => Err(ProtocolError::MissingData),
+        false => {
+            let mut split = bytes.splitn(2, |b| b == &b'\n');
+            let header = split.next().ok_or(ProtocolError::MissingHeaders)?;
+
+            // Try to parse the headers to determine if they are valid. This is a good sanity
+            // check to determine if our headers extraction is working properly.
+            serde_json::from_slice::<Headers>(header).map_err(ProtocolError::InvalidHeaders)?;
+
+            let body = match split.next() {
+                Some(b"") | None => return Err(ProtocolError::MissingBody),
+                Some(body) => body,
+            };
+
+            Ok((header, body))
+        }
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Deserialize)]
+struct Headers {
+    segment_id: u16,
+}
+
+#[derive(Debug)]
+pub enum ProtocolError {
+    EncodingError(Error),
+    InvalidHeaders(SError),
+    MissingData,
+    MissingHeaders,
+    MissingBody,
+}
+
+impl Display for ProtocolError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ProtocolError::InvalidHeaders(e) => {
+                write!(f, "recording headers could not be parsed {e:?}")
+            }
+            ProtocolError::MissingData => write!(f, "no recording found"),
+            ProtocolError::MissingHeaders => write!(f, "no recording headers found"),
+            ProtocolError::MissingBody => write!(f, "not recording body found"),
+            ProtocolError::EncodingError(e) => write!(f, "{e:?}"),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::recording::protocol::{deserialize, serialize, ProtocolError};
+    use serde_json::Value;
+
+    #[test]
+    fn test_deserialize_message() {
+        // Contains a 16 byte header followed by a new line character.
+        let payload: &[u8] = &[
+            123, 34, 115, 101, 103, 109, 101, 110, 116, 95, 105, 100, 34, 58, 51, 125, 10, 120,
+        ];
+
+        let result = deserialize(payload, 9999999999);
+        assert!(result.is_ok());
```

Comment:
Instead of asserting `is_ok()` here, just unwrap. This will print the error and fail the test automatically.